swarm/swarm.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/deis/deis/tests/utils"
+)
+
+//Response struct
+type Response struct {
+	ClientURL string `json:"clientURL"`
+	Name      string `json:"name"`
+	PeerURL   string `json:"peerURL"`
+	State     string `json:"state"`
+}
+
+const (
+	swarmpath               = "/deis/scheduler/swarm/node"
+	swarmetcd               = "/deis/scheduler/swarm/host"
+	etcdport                = "4001"
+	timeout   time.Duration = 3 * time.Second
+	ttl       time.Duration = timeout * 2
+)
+
+func run(cmd string) {
+	var cmdBuf bytes.Buffer
+	tmpl := template.Must(template.New("cmd").Parse(cmd))
+	if err := tmpl.Execute(&cmdBuf, nil); err != nil {
+		log.Fatal(err)
+	}
+	cmdString := cmdBuf.String()
+	fmt.Println(cmdString)
+	var cmdl *exec.Cmd
+	cmdl = exec.Command("sh", "-c", cmdString)
+	if _, _, err := utils.RunCommandWithStdoutStderr(cmdl); err != nil {
+		log.Fatal(err)
+	} else {
+		fmt.Println("ok")
+	}
+}
+
+func getleaderHost() string {
+	var response []Response
+	var host string
+	client := &http.Client{}
+	resp, _ := client.Get("http://" + os.Getenv("HOST") + ":7001/v2/admin/machines")
+	defer resp.Body.Close()
+	body, _ := ioutil.ReadAll(resp.Body)
+	json.Unmarshal(body, &response)
+	for _, node := range response {
+		if node.State == "leader" {
+			host = strings.Split(node.ClientURL, "//")[1]
+		}
+	}
+	return host
+}
+
+func publishService(client *etcd.Client, host string, ttl uint64) {
+	for {
+		setEtcd(client, swarmetcd, host, ttl)
+		time.Sleep(timeout)
+	}
+}
+
+func setEtcd(client *etcd.Client, key, value string, ttl uint64) {
+	_, err := client.Set(key, value, ttl)
+	if err != nil {
+		log.Println(err)
+	}
+}
+func main() {
+	etcdproto := "etcd://" + getleaderHost() + swarmpath
```

Comment:
I just tried swarm join with etcd and I think there is bug in the latest swarm image from docker repository.Only the last syntax works with etcd protocol
I have tried swarm join with the following  etcd protocol syntax only the last syntax worked. 
```
docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100/mod/swarm
time="2015-05-05T16:29:10Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]" 
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100,172.17.8.101,172.17.8.102,172.17.8.103/mod/swarm
time="2015-05-05T16:33:36Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]"
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100:4001,172.17.8.101:4001,172.17.8.102:4001,172.17.8.103:4001/mod/swarm
time="2015-05-05T16:40:12Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]"
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100:4001/mod/swarm
time="2015-05-05T16:32:44Z" level=info msg="Registering on the discovery service every 25 seconds..." addr="172.17.8.103:2375" discovery="etcd://172.17.8.100:4001/mod/swarm" 
```

172.17.8.100 is the leader host.
I'll open an issue with docker/swarm